README
									README
##-----------------------------------------------------------------------------##
# GEOCONVERTOR JAVA API version BETA - Converts coordinate latitude/longitude.
# https://github.com/razec/geoconvertor


News: (For news, check CHANGELOG file) and http://code.google.com/p/geoconvertor/

What is this?

The geoconvertor it's API JAVA created to convert coordinates using complex geodesic equation. 

Supporting three conversions:

	Decimal Coordinate to UTM
		Easily convert between latitude/longitude, Universal Transverse Mercator (UTM) using DATUM: SAD69 and WGS-84.
	DMS (Degrees, Minutes, Seconds) to Decimal Degree
	
	Decimal Degree to DMS (Degrees, Minutes, Seconds)

Pre-requisites:

Java:
- Java Runtime Environment Version 5.0

How to install?

1) Adding geoconvertor-v0.x.jar api to the project.

2) Import
import org.geoconvertor.datum.Datum;
import org.geoconvertor.utm.GEOConvertor;

Construct an object:

GEOConvertor gcv = new GEOConvertor(); 

Call the methods:

geoToUTM(decimal_latitude, decimal_longitude, datum, central_meridian);
decToDMS(decimal_coordinate);
dmsToDec(dms_coordinate); 

More Information: https://github.com/razec/geoconvertor/wiki




CHANGELOG
0.1
- Initial Version.
- Bug: Necessary more precision in the coordinate.


teste/test/org/geo/spirit/Datum.java
package test.org.geo.spirit;

public enum Datum {
	SAD69(6378160.0000, 6356774.7192), WGS84(6378137.0000, 6356752.31424518);
	double MAJOR_AXIS;
	double MINOR_AXIS;
	double FLATTENING;

	private Datum(double MAJOR_AXIS, double MINOR_AXIS) {
		this.MAJOR_AXIS = MAJOR_AXIS;
		this.MINOR_AXIS = MINOR_AXIS;
		FLATTENING = Ellipsoid.flattening(MAJOR_AXIS, MINOR_AXIS);
	}
}


teste/test/org/geo/spirit/CalcJ2me.java
package test.org.geo.spirit;

public class CalcJ2me implements Calc {
	public double pow(double base, double expoente) {
		return 0;
	}

}


teste/test/org/geo/spirit/CalcJ2se.java
package test.org.geo.spirit;

public class CalcJ2se implements Calc {
	public double pow(double base, double expoente) {
		return Math.pow(base, expoente);
	}
}


src/org/geoconvertor/me/IConstants.java
package org.geoconvertor.me;
/**
 * @(#)IConstants.java	0.1 07/05/2010 
 * 
 * This program is free software; you can redistribute it and/or                
 * modify it under the terms of the GNU General Public License                  
 * as published by the Free Software Foundation; either version 3               
 * of the License, or (at your option) any later version.  
 * 
 * <p>
 * This class is part of the GEOConvertor package. Visit the <a
 * href="http://code.google.com/p/geoconvertor/">GEOConvertor</a> website for more
 * information.
 * </p>
 * 
 * <p>
 *  An interface to implement constant values. 
 * </p>
 * 
 * @author Mario C. Ponciano | a.k.a Razec - mrazec@gmail.com
 * @version 0.1 - 07/05/2010
 * @since 0.1
 * @see org.geoconvertor.utm.GEOConvertor
 */
public interface IConstants {
	/**
	 * Sine of one second of arc = pi / (180*60*60) = 4.8481368x10 (exp.) -6.
	 * 
	 * The value of this constant is {@SIN_ONE_SECOND}.
	 */
	public static final double SIN_ONE_SECOND = 0.00000484813681108; //Sin 1"
	/**
	 * Indicates that the Scale Factor {@K0}.
	 */
	public static final double K0 = 0.999995; //Scale Factor
}


src/org/geoconvertor/me/AbstractEllipsoid.java
package org.geoconvertor.me;
/**
 *  @(#)AbstractEllipsoid.java	0.1 07/05/2010 
 * 
 * This program is free software; you can redistribute it and/or                
 * modify it under the terms of the GNU General Public License                  
 * as published by the Free Software Foundation; either version 3               
 * of the License, or (at your option) any later version.  
 * 
 * <p>
 * This class is part of the GEOConvertor package. Visit the <a
 * href="http://code.google.com/p/geoconvertor/">GEOConvertor</a> website for more
 * information.
 * </p>
 * 
 * <p>
 *	Abstract Class ellipsoid calculate.
 *<p>
 * -flatattening earth
 * </p>
 *<p>
 * -eccentricity (E) and (E')
 * </p>
 *<p>
 * -Big N
 * </p>
 *<p>
 * -Delta Long 
 * </p>
 *<p>
 * -P
 * </p>
 * 
 * </p>
 * 
 * 
 * @author Mario C. Ponciano | a.k.a Razec - mrazec@gmail.com
 * @version 0.1 - 07/05/2010
 * @since 0.1
 */

public abstract class AbstractEllipsoid extends AbstractCoefficient {
	/**
	 *Method Calculate flattening earth.
	 * 
	 * @param majorAxis
	 * @param minorAxis
	 * @return
	 */
	public double flattening(double majorAxis, double minorAxis) {
		return (majorAxis - minorAxis) / majorAxis;
	}

	/**
	 * Method is the eccentricity of the earth's elliptical.
	 * 
	 * @param majorAxis
	 * @param minorAxis
	 * @return - the double eccentricity one.
	 */
	protected double eccentricityOne(double majorAxis, double minorAxis){
		return (Math.sqrt((majorAxis*majorAxis) - (minorAxis*minorAxis))) / majorAxis;
	}

	/**
	 * 
	 * 
	 * @param majorAxis
	 * @param minorAxis
	 * @return
	 */
	protected double eccentricityTwo(double majorAxis, double minorAxis){
		return (Math.sqrt((majorAxis * majorAxis) - (minorAxis*minorAxis))) / minorAxis;
	}

	/**
	 * 
	 * Calculate the big normal, ellipse-rotation method used to measure around minorAxis.
	 * 
	 * @param majorAxis
	 * @param eccentricityOne
	 * @param latRadian
	 * @return
	 */
	protected double bigN(double majorAxis, double eccentricityOne, double latRadian){
		return majorAxis / Math.sqrt((1 - (eccentricityOne * eccentricityOne)* (Math.sin(latRadian)*Math.sin(latRadian))));
	}

	protected double deltaLong(double lngDecimal, int meridianCentral){
		return (lngDecimal - meridianCentral) * 3600;
	}

	protected double p(double deltaLong){
		return  0.0001 * deltaLong;
	}

}


src/org/geoconvertor/me/AbstractCoefficient.java
package org.geoconvertor.me;

/**
 * @(#)AbstractCoefficient.java	0.1 07/05/2010 
 * 
 * This program is free software; you can redistribute it and/or                
 * modify it under the terms of the GNU General Public License                  
 * as published by the Free Software Foundation; either version 3               
 * of the License, or (at your option) any later version.
 * 
 * <p>
 * This class is part of the GEOConvertor package. Visit the <a
 * href="http://code.google.com/p/geoconvertor/">GEOConvertor</a> website for more
 * information.
 * </p>
 * 
 * <p>
 * Abstract Class determination of coefficient.
 * <p>
 * Coefficients: A, B, C, D, E, F
 * </p>
 * <p>
 * Coefficients to (North) and (East): I, S, IIP2, IIP4, IVP, VP3, A'6P6, B'5P5.
 * </p>
 * </p>
 * 
 * 
 * @author Mario C. Ponciano | a.k.a Razec - mrazec@gmail.com
 * @version 0.1 - 07/05/2010
 * @since 0.1
 */
import org.geoconvertor.utm.IConstants;

public abstract class AbstractCoefficient {

	 public final static double getPow(double base, int exp) {
		double ret = base;
		int i = 1;
		while (i < exp) {
			ret = (base * ret);
			i++;
		}
		return ret;
	}

	protected double aCoeff(double eccentricityOne) {
		return (1 + 3 / 4 * getPow(eccentricityOne, 2) + 45 / 64
				* getPow(eccentricityOne, 4) + 175 / 256
				* getPow(eccentricityOne, 6) + 11025 / 16384
				* getPow(eccentricityOne, 8) + 43656 / 65536 * getPow(
				eccentricityOne, 10));
	}

	protected double bCoeff(double eccentricityOne) {
		return (3 / 4 * getPow(eccentricityOne, 2) + 15 / 16
				* getPow(eccentricityOne, 4) + 525 / 512
				* getPow(eccentricityOne, 6) + 2205 / 2048
				* getPow(eccentricityOne, 8) + 72765 / 65536 * getPow(
				eccentricityOne, 10));
	}

	protected double cCoeff(double eccentricityOne) {
		return (15 / 64 * getPow(eccentricityOne, 4) + 105 / 256
				* getPow(eccentricityOne, 6) + 2205 / 4096
				* getPow(eccentricityOne, 8) + 10395 / 16384 * getPow(
				eccentricityOne, 10));
	}

	protected double dCoeff(double eccentricityOne) {
		return (35 / 512 * getPow(eccentricityOne, 6) + 315 / 2048
				* getPow(eccentricityOne, 8) + 31185 / 131072 * getPow(
				eccentricityOne, 10));
	}

	protected double eCoeff(double eccentricityOne) {
		return (315 / 16384 * getPow(eccentricityOne, 8) + 3465 / 65536 * getPow(
				eccentricityOne, 10));
	}

	protected double fCoeff(double eccentricityOne) {
		return (639 / 131072 * getPow(eccentricityOne, 10));
	}

	/**
	 * Method calculate (S) the meridional arc, the distance along the earth's
	 * surface from the equator.
	 * 
	 * @param majorAxis
	 * @param lat
	 * @param eccentricityOne
	 * @param aCoeff
	 * @param bCoeff
	 * @param cCoeff
	 * @param dCoeff
	 * @param eCoeff
	 * @param fCoeff
	 * @return
	 */
	protected double sCoeff(double majorAxis, double lat,
			double eccentricityOne, double aCoeff, double bCoeff,
			double cCoeff, double dCoeff, double eCoeff, double fCoeff) {
		return ((majorAxis) * (1 - getPow(eccentricityOne, 2)) * (aCoeff
				* (lat * Math.PI / 180) - 1 / 2 * bCoeff
				* Math.sin(2 * Math.toRadians(lat)) + 1 / 4 * cCoeff
				* Math.sin(4 * Math.toRadians(lat)) - 1 / 6 * dCoeff
				* Math.sin(6 * Math.toRadians(lat)) + 1 / 8 * eCoeff
				* Math.sin(8 * Math.toRadians(lat)) - 1 / 10 * fCoeff
				* Math.sin(10 * Math.toRadians(lat))));
	}

	protected double i(double sCoeff) {
		return IConstants.K0 * sCoeff;
	}

	protected double iIP2(double n, double lat, double p) {
		return (n * Math.sin(Math.toRadians(lat))
				* Math.cos(Math.toRadians(lat))
				* getPow(IConstants.SIN_ONE_SECOND, 2) / 2
				* getPow(IConstants.K0 * 10, 8) * getPow(p, 2));
	}

	protected double iIP4(double n, double lat, double eccentricityTwo, double p) {
		return (getPow(IConstants.SIN_ONE_SECOND, 4) * n
				* Math.sin(Math.toRadians(lat))
				* getPow(Math.cos(Math.toRadians(lat)), 3) / 24)
				* (5 - (getPow(Math.tan(Math.toRadians(lat)), 2)) + 9
						* getPow(eccentricityTwo, 2)
						* getPow(Math.cos(Math.toRadians(lat)), 2) + 4
						* getPow(eccentricityTwo, 4)
						* getPow(Math.cos(Math.toRadians(lat)), 4))
				* IConstants.K0 * getPow(10, 16) * getPow(p, 4);
	}

	protected double iVP(double n, double lat, double p) {
		return (n * Math.cos(Math.toRadians(lat)) * IConstants.SIN_ONE_SECOND
				* IConstants.K0 * getPow(10, 4))
				* p;
	}

	protected double VP3(double n, double lat, double p, double eccentricityTwo) {
		return (getPow(IConstants.SIN_ONE_SECOND, 3) * n
				* getPow(Math.cos(Math.toRadians(lat)), 3) / 6)
				* (1 - (getPow(Math.tan(Math.toRadians(lat)), 2) + getPow(
						eccentricityTwo, 2)
						* getPow(Math.cos(Math.toRadians(lat)), 2))
						* IConstants.K0) * getPow(10, 12) * getPow(p, 3);
	}

	protected double a6P6(double n, double lat, double eccentricityTwo, double p) {
		return ((getPow(IConstants.SIN_ONE_SECOND, 6) * n
				* Math.sin(Math.toRadians(lat))
				* getPow(Math.cos(Math.toRadians(lat)), 5) / 720)
				* (61 - 58 * getPow(Math.tan(Math.toRadians(lat)), 2)
						+ getPow(Math.tan(Math.toRadians(lat)), 4) + 270
						* getPow(eccentricityTwo, 2)
						* getPow(Math.cos(Math.toRadians(lat)), 2) - 330
						* getPow(eccentricityTwo, 2)
						* getPow(Math.sin(Math.toRadians(lat)), 2)
						* IConstants.K0) * getPow(10, 24))
				* getPow(p, 6);
	}

	protected double b5P5(double n, double lat, double eccentricityTwo, double p) {
		return (getPow(IConstants.SIN_ONE_SECOND, 5) * n
				* getPow(Math.cos(Math.toRadians(lat)), 5) / 120)
				* (5 - 18 * getPow(Math.tan(Math.toRadians(lat)), 2)
						+ getPow(Math.tan(Math.toRadians(lat)), 4) + 14
						* getPow(eccentricityTwo, 2)
						* getPow(Math.cos(Math.toRadians(lat)), 2) - 58
						* getPow(eccentricityTwo, 2)
						* getPow(Math.sin(Math.toRadians(lat)), 2))
				* IConstants.K0 * getPow(10, 20) * getPow(p, 5);
	}

}


src/org/geoconvertor/me/SAD69.java
package org.geoconvertor.me;

/**
 * @(#)SAD69.java	0.1 07/05/2010 
 * 
 * This program is free software; you can redistribute it and/or                
 * modify it under the terms of the GNU General Public License                  
 * as published by the Free Software Foundation; either version 3               
 * of the License, or (at your option) any later version.  
 * 
 * <p>
 * This class is part of the GEOConvertor package. Visit the <a
 * href="http://code.google.com/p/geoconvertor/">GEOConvertor</a> website for more
 * information.
 * </p>
 * 
 * <p>
 * Constant Class defining the South American 1969 reference ellipsoid.
 * </p>
 * 
 * 
 * @author Mario C. Ponciano | a.k.a Razec - mrazec@gmail.com
 * @version 0.1 - 07/05/2010
 * @since 0.1
 */
public class SAD69 {
	/**
	 * Indicates that the semi-major axis {@MAJOR_AXIS}
	 */
	public static final double MAJOR_AXIS = 6378160.0000;
	/**
	 * Indicates that the semi-minor axis {@MINOR_AXIS}
	 */
	public static final double MINOR_AXIS= 6356774.7192;

}
